src/internal/common/prefix_matcher.go

```diff
@@ -0,0 +1,75 @@
+package common
+
+import (
+	"strings"
+
+	"github.com/alcionai/clues"
+)
+
+type PrefixMatcher[T any] interface {
+	Add(key string, value T) error
+	Get(key string) (T, bool)
+	LongestPrefix(key string) (string, T, bool)
+	Map() map[string]T
+	Empty() bool
+}
+
+type prefixMatcher[T any] struct {
+	data map[string]T
+}
+
+func (m *prefixMatcher[T]) Add(key string, value T) error {
+	if _, ok := m.data[key]; ok {
+		return clues.New("entry already exists").With("key", key)
+	}
```

Comment:
This feels overbearing.  It's not standard for a map implementations to error on collision.  Is an error necessary?  It seems acceptable to allow users to overwrite keys.